### PR TITLE
Correct order of ossl_condvar_signal in quic_multistream_test

### LINKED ISSUE #22588: [3.2.0-beta1] quicapitest hangs for NonStop thread (SPT) builds
When testing NonStop builds for 3.2.0-beta1, `quicapitest` (and sometimes `quic_tserver_test`) hangs waiting for initial input. This did not occur with 3.2.0-alpha2 but this may be transient.

Stopping the hung processes to not cause the tests to fail, which is odd itself.

This only applies to pthread builds using the SPT model. PUT is fine, but it is much faster implementation so any transient problems may get masked. Unthreaded builds work consistently fine.

Config dump:
```
Command line (with current working directory = .):

    perl ./Configure nonstop-nsx_spt_floss --prefix=/usr/local-ssl3.2 --openssldir=/usr/local-ssl3.2/ssl --with-rand-seed=rdcpu threads enable-fips -D_REENTRANT

Perl information:

    perl
    5.30.3 for NSV-D-nonstop_kernel

Enabled features:

    acvp-tests
    apps
    argon2
    aria
    async
    autoalginit
    autoerrinit
    autoload-config
    bf
    blake2
    bulk
    cached-fetch
    camellia
    capieng
    cast
    chacha
    cmac
    cmp
    cms
    comp
    ct
    default-thread-pool
    deprecated
    des
    dgram
    dh
    docs
    dsa
    dso
    dtls
    dynamic-engine
    ec
    ec2m
    ecdh
    ecdsa
    ecx
    egd
    engine
    err
    filenames
    fips
    fips-securitychecks
    gost
    http
    idea
    legacy
    loadereng
    md4
    mdc2
    module
    multiblock
    nextprotoneg
    ocb
    ocsp
    padlockeng
    pic
    pinshared
    poly1305
    posix-io
    psk
    quic
    rc2
    rc4
    rdrand
    rfc3779
    rmd160
    scrypt
    secure-memory
    seed
    shared
    siphash
    siv
    sm2
    sm2-precomp
    sm3
    sm4
    sock
    srp
    srtp
    sse2
    ssl
    ssl-trace
    static-engine
    stdio
    tests
    thread-pool
    threads
    tls
    ts
    ui-console
    whirlpool
    tls1
    tls1-method
    tls1_1
    tls1_1-method
    tls1_2
    tls1_2-method
    tls1_3
    dtls1
    dtls1-method
    dtls1_2
    dtls1_2-method

Disabled features:

    afalgeng            [not-linux]      OPENSSL_NO_AFALGENG
    asan                [default]        OPENSSL_NO_ASAN
    asm                 [no asm_arch]    OPENSSL_NO_ASM
    brotli              [default]        OPENSSL_NO_BROTLI
    brotli-dynamic      [default]        OPENSSL_NO_BROTLI_DYNAMIC
    buildtest-c++       [default]
    winstore            [not-windows]    OPENSSL_NO_WINSTORE
    crypto-mdebug       [default]        OPENSSL_NO_CRYPTO_MDEBUG
    devcryptoeng        [default]        OPENSSL_NO_DEVCRYPTOENG
    ec_nistp_64_gcc_128 [default]        OPENSSL_NO_EC_NISTP_64_GCC_128
    external-tests      [default]        OPENSSL_NO_EXTERNAL_TESTS
    fuzz-afl            [default]        OPENSSL_NO_FUZZ_AFL
    fuzz-libfuzzer      [default]        OPENSSL_NO_FUZZ_LIBFUZZER
    ktls                [default]        OPENSSL_NO_KTLS
    makedepend          [unavailable]
    md2                 [default]        OPENSSL_NO_MD2 (skip crypto/md2)
    msan                [default]        OPENSSL_NO_MSAN
    rc5                 [default]        OPENSSL_NO_RC5 (skip crypto/rc5)
    sctp                [default]        OPENSSL_NO_SCTP
    tfo                 [default]        OPENSSL_NO_TFO
    trace               [default]        OPENSSL_NO_TRACE
    ubsan               [default]        OPENSSL_NO_UBSAN
    unit-test           [default]        OPENSSL_NO_UNIT_TEST
    uplink              [no uplink_arch] OPENSSL_NO_UPLINK
    weak-ssl-ciphers    [default]        OPENSSL_NO_WEAK_SSL_CIPHERS
    zlib                [default]        OPENSSL_NO_ZLIB
    zlib-dynamic        [default]        OPENSSL_NO_ZLIB_DYNAMIC
    zstd                [default]        OPENSSL_NO_ZSTD
    zstd-dynamic        [default]        OPENSSL_NO_ZSTD_DYNAMIC
    ssl3                [default]        OPENSSL_NO_SSL3
    ssl3-method         [default]        OPENSSL_NO_SSL3_METHOD

Config target attributes:

    AR => "ar",
    ARFLAGS => "qc",
    CC => "c99",
    HASHBANGPERL => "/usr/bin/env perl",
    RANLIB => "ranlib",
    RC => "windres",
    bn_ops => "THIRTY_TWO_BIT",
    build_file => "Makefile",
    build_scheme => [ "unified", "unix" ],
    cflags => "-g -O2 -Wextensions -Wnowarn=203,220,272,734,770,1506 -Wbuild_neutral_library -Wverbose -Wtarget=tns/x -Wsystype=oss -Wilp32 -WIEEE_float",
    cppflags => "",
    defines => [ "OPENSSL_BUILDING_OPENSSL", "OPENSSL_VPROC=\$(OPENSSL_VPROC)", "_XOPEN_SOURCE", "_XOPEN_SOURCE_EXTENDED=1", "_TANDEM_SOURCE", "__NSK_OPTIONAL_TYPES__", "B_ENDIAN", "OPENSSL_TANDEM_FLOSS", "_SPT_MODEL_", "_REENTRANT", "_ENABLE_FLOSS_THREADS" ],
    disable => [  ],
    dso_scheme => "DLFCN",
    enable => [ "egd" ],
    ex_libs => "-lrld -lfloss -lspt",
    includes => [ "/usr/local/include" ],
    lflags => "-Wxld=\"-set systype oss\" -Wxld=\"-set floattype ieee_float\"",
    lib_cflags => "",
    lib_cppflags => "",
    lib_defines => [  ],
    module_cflags => "",
    module_cxxflags => undef,
    module_ldflags => "-Wshared",
    multibin => "-spt",
    multilib => "-spt",
    perl => "/usr/bin/perl",
    perl_platform => "Unix",
    shared_argfileflag => "-Wxld_obey=",
    shared_cflag => "",
    shared_defflag => "-Wxld_obey=",
    shared_defines => [  ],
    shared_extension => ".so",
    shared_ldflag => "-Wshared",
    shared_rcflag => "",
    shared_target => "nonstop-shared",
    sys_id => "TANDEM",
    thread_defines => [  ],
    thread_scheme => "(unknown)",
    unistd => "<unistd.h>",

Recorded environment:

    AR =
    BUILDFILE =
    CC =
    CFLAGS =
    CPPFLAGS =
    CROSS_COMPILE =
    CXX =
    CXXFLAGS =
    HASHBANGPERL =
    LDFLAGS =
    LDLIBS =
    OPENSSL_LOCAL_CONFIG_DIR =
    PERL =
    RANLIB =
    RC =
    RCFLAGS =
    WINDRES =
    __CNF_CFLAGS =
    __CNF_CPPDEFINES =
    __CNF_CPPFLAGS =
    __CNF_CPPINCLUDES =
    __CNF_CXXFLAGS =
    __CNF_LDFLAGS =
    __CNF_LDLIBS =

Makevars:

    AR              = ar
    ARFLAGS         = qc
    ASFLAGS         =
    CC              = c99
    CFLAGS          =
    CPPDEFINES      = _REENTRANT
    CPPFLAGS        =
    CPPINCLUDES     =
    CXXFLAGS        =
    HASHBANGPERL    = /usr/bin/env perl
    LDFLAGS         =
    LDLIBS          =
    PERL            = perl
    RANLIB          = ranlib
    RC              = windres
    RCFLAGS         =

NOTE: These variables only represent the configuration view.  The build file
template may have processed these variables further, please have a look at the
build file for more exact data:
    Makefile

build file:

    Makefile

build file templates:

    Configurations/common0.tmpl
    Configurations/unix-Makefile.tmpl
```

### PR DESCRIPTION
**Pull request by nhorman**
quic_multistream test was issuing a signal on a condvar after dropping the corresponding mutex, not before, leading to potential race conditions in the reading of the associated data

Related to #22588

### CODE REVIEW

**Review by hlandau — APPROVED** [review]

**Comment by mattcaswell** [comment]
LGTM...pending this coming out of draft before approval.

**Comment by nhorman** [comment]
I think this is probably worth merging, even though it doesn't fix the nonstop problem yet, as locking shared data (fake_now), and correcting the mutex/signal order is the right thing to do in terms of programming patterns

File: test/helpers/quictestlib.c
Comment (hlandau) [inline test/helpers/quictestlib.c:86]: `= ossl_time_zero()` (don't access the internal structure of OSSL_TIME directly)

File: test/helpers/quictestlib.c
Comment (hlandau) [inline test/helpers/quictestlib.c:293]: What if this fails? Subsequently we will then call CRYPTO_THREAD_write_lock(NULL)

File: test/helpers/quictestlib.c
Comment (hlandau) [inline test/helpers/quictestlib.c:417]: We should probably move all this locking into `qtest_get_time` and `qtest_set_time` to go along with the existing `qtest_add_time`.

**Review by hlandau — CHANGES_REQUESTED** [review]

**Comment by hlandau** [comment]
Few comments, otherwise looks good

File: test/helpers/quictestlib.c
Comment (paulidale) [reply test/helpers/quictestlib.c:86]: The reason being that the internal format is subject to change.

File: test/helpers/quictestlib.c
Comment (nhorman) [reply test/helpers/quictestlib.c:432]: updated to create qtest_reset_time and qtest_get_time, to isolate the locking

File: test/helpers/quictestlib.c
Comment (nhorman) [reply test/helpers/quictestlib.c:289]: Updated to fail qtest_create_quic_objects  if the lock creation falis

File: test/helpers/quictestlib.c
Comment (paulidale) [inline test/helpers/quictestlib.c:359]: Write lock here.

**Review by hlandau — APPROVED** [review]

**Review by mattcaswell — APPROVED** [review]

**Comment by levitte** [comment]
This might fix the issues I've noticed on VMS.  I'll have a look this evening

**Review by paulidale — APPROVED** [review]

File: test/helpers/quictestlib.c
Comment (t8m) [inline test/helpers/quictestlib.c:364]: Is this racy? Could you just ossl_time_add() inside the write lock?

Alternatively you can see this function as something that should be never used in racy contexts and drop the lock calls completely here. Using this in a racy context would mean that we need to have multiple fake_now storages per use actually because otherwise one thread could reset the fake time for another innocent thread and the time would go backwards.

File: test/helpers/quictestlib.c
Comment (t8m) [inline test/helpers/quictestlib.c:436]: The curly brackets are superfluous.

File: test/helpers/quictestlib.c
Comment (t8m) [inline test/helpers/quictestlib.c:446]: Same here about the curly brackets. Please see the coding style.

**Review by t8m — CHANGES_REQUESTED** [review]

File: test/helpers/quictestlib.c
Comment (t-j-h) [reply test/helpers/quictestlib.c:446]: The coding style does not prevent (or require) braces in this context ...

Comment (t8m) [reply test/helpers/quictestlib.c:446]: It says:

Do not unnecessarily use braces around a single statement:

    if (condition)
        action();

and

    if (condition)
        do_this();
    else
        do_that();

File: test/helpers/quictestlib.c
Comment (nhorman) [reply test/helpers/quictestlib.c:364]: I don't think this is racy.  qtest_add_time retakes the global lock, so there aren't any read/write races strictly speaking.  It is theoretically possible that another thread might read the zero time, but given that this is during setup, there should be no other threads running

File: test/helpers/quictestlib.c
Comment (t-j-h) [reply test/helpers/quictestlib.c]: You are correct - I remembered the debates on this - not the outcome - which is to not use them around a single statement.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
As the last fixup is only a trivial code style fix no re-approvals are IMO needed.

**Comment by t8m** [comment]
Merged to the master and 3.2 branches. Thank you.

**Comment by levitte** [comment]
Unfortunately, this didn't fix the VMS test issue either...  [sigh]

**Pull request closed without merging**
